File: android/app/src/main/kotlin/com/amazonaws/services/chime/flutterdemo/RealtimeObserver.kt
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: MIT-0
 */
 
package com.amazonaws.services.chime.flutterdemo

import com.amazonaws.services.chime.sdk.meetings.realtime.RealtimeObserver
import com.amazonaws.services.chime.sdk.meetings.audiovideo.AttendeeInfo
import com.amazonaws.services.chime.sdk.meetings.audiovideo.SignalUpdate
import com.amazonaws.services.chime.sdk.meetings.audiovideo.VolumeUpdate

class RealtimeObserver(val methodChannel: MethodChannelCoordinator) : RealtimeObserver {

    override fun onAttendeesDropped(attendeeInfo: Array<AttendeeInfo>) {
        for (currentAttendeeInfo in attendeeInfo) {
            methodChannel.callFlutterMethod(
                MethodCall.drop,
                attendeeInfoToMap(currentAttendeeInfo)
            )
        }
    }

    override fun onAttendeesJoined(attendeeInfo: Array<AttendeeInfo>) {
        for (currentAttendeeInfo in attendeeInfo) {
            methodChannel.callFlutterMethod(
                MethodCall.join,
                attendeeInfoToMap(currentAttendeeInfo)
            )
        }
    }

    override fun onAttendeesLeft(attendeeInfo: Array<AttendeeInfo>) {
        for (currentAttendeeInfo in attendeeInfo) {
            methodChannel.callFlutterMethod(
                MethodCall.leave,
                attendeeInfoToMap(currentAttendeeInfo)
            )
        }
    }

    override fun onAttendeesMuted(attendeeInfo: Array<AttendeeInfo>) {
        for (currentAttendeeInfo in attendeeInfo) {
            methodChannel.callFlutterMethod(MethodCall.mute, attendeeInfoToMap(currentAttendeeInfo))
        }
    }

    override fun onAttendeesUnmuted(attendeeInfo: Array<AttendeeInfo>) {
        for (currentAttendeeInfo in attendeeInfo) {
            methodChannel.callFlutterMethod(
                MethodCall.unmute,
                attendeeInfoToMap(currentAttendeeInfo)
            )
        }
    }

    override fun onSignalStrengthChanged(signalUpdates: Array<SignalUpdate>) {
        // Out of Scope
    }

    override fun onVolumeChanged(volumeUpdates: Array<VolumeUpdate>) {
        // Out of Scope
    }

    private fun attendeeInfoToMap(attendee: AttendeeInfo): Map<String, Any?> {
        return mapOf(
            "attendeeId" to attendee.attendeeId,
            "externalUserId" to attendee.externalUserId
        )
    }
}

File: ios/Runner/RealtimeObserver.swift
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: MIT-0
 */
 
import AmazonChimeSDK
import AmazonChimeSDKMedia
import AVFoundation
import Flutter
import Foundation
import UIKit

class MyRealtimeObserver: RealtimeObserver {
    weak var methodChannel: MethodChannelCoordinator?

    init(withMethodChannel methodChannel: MethodChannelCoordinator) {
        self.methodChannel = methodChannel
    }

    func volumeDidChange(volumeUpdates: [VolumeUpdate]) {
        // Out of scope
    }

    func signalStrengthDidChange(signalUpdates: [SignalUpdate]) {
        // Out of scope
    }

    func attendeesDidJoin(attendeeInfo: [AttendeeInfo]) {
        for currentAttendeeInfo in attendeeInfo {
            methodChannel?.callFlutterMethod(method: .join, args: attendeeInfoToDictionary(attendeeInfo: currentAttendeeInfo))
        }
    }

    func attendeesDidLeave(attendeeInfo: [AttendeeInfo]) {
        for currentAttendeeInfo in attendeeInfo {
            methodChannel?.callFlutterMethod(method: .leave, args: attendeeInfoToDictionary(attendeeInfo: currentAttendeeInfo))
        }
    }

    func attendeesDidDrop(attendeeInfo: [AttendeeInfo]) {
        for currentAttendeeInfo in attendeeInfo {
            methodChannel?.callFlutterMethod(method: .drop, args: attendeeInfoToDictionary(attendeeInfo: currentAttendeeInfo))
        }
    }

    func attendeesDidMute(attendeeInfo: [AttendeeInfo]) {
        for currentAttendeeInfo in attendeeInfo {
            methodChannel?.callFlutterMethod(method: .mute, args: attendeeInfoToDictionary(attendeeInfo: currentAttendeeInfo))
        }
    }

    func attendeesDidUnmute(attendeeInfo: [AttendeeInfo]) {
        for currentAttendeeInfo in attendeeInfo {
            methodChannel?.callFlutterMethod(method: .unmute, args: attendeeInfoToDictionary(attendeeInfo: currentAttendeeInfo))
        }
    }

    private func attendeeInfoToDictionary(attendeeInfo: AttendeeInfo) -> [String: String] {
        return [
            "attendeeId": attendeeInfo.attendeeId,
            "externalUserId": attendeeInfo.externalUserId
        ]
    }
}
